aws/resource_aws_kinesis_video_stream.go

```diff
@@ -0,0 +1,266 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/kinesisvideo"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+
+	"github.com/terraform-providers/terraform-provider-aws/aws/internal/keyvaluetags"
+)
+
+func resourceAwsKinesisVideoStream() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsKinesisVideoStreamCreate,
+		Read:   resourceAwsKinesisVideoStreamRead,
+		Update: resourceAwsKinesisVideoStreamUpdate,
+		Delete: resourceAwsKinesisVideoStreamDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(5 * time.Minute),
+			Update: schema.DefaultTimeout(120 * time.Minute),
+			Delete: schema.DefaultTimeout(120 * time.Minute),
+		},
+
+		Schema: map[string]*schema.Schema{
+			"arn": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"data_retention_in_hours": {
+				Type:     schema.TypeInt,
+				Optional: true,
+				ForceNew: true,
+				Default:  0,
+			},
+
+			"device_name": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: validateAwsKinesisVideoStreamDeviceName,
+			},
+
+			"kms_key_id": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+				ForceNew: true,
+			},
+
+			"media_type": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"creation_time": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"version": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"tags": tagsSchema(),
+		},
+	}
+}
+
+func resourceAwsKinesisVideoStreamCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).kinesisvideoconn
+
+	createOpts := &kinesisvideo.CreateStreamInput{
+		StreamName:           aws.String(d.Get("name").(string)),
+		DataRetentionInHours: aws.Int64(int64(d.Get("data_retention_in_hours").(int))),
+		Tags:                 keyvaluetags.New(d.Get("tags").(map[string]interface{})).IgnoreAws().KinesisvideoTags(),
```

Comment:
It appears the Kinesis Video SDK is sensitive to sending a zero-length `Tags` field instead of `nil`:

```
--- FAIL: TestAccAWSKinesisVideoStream_basic (4.36s)
    testing.go:654: Step 0 error: errors during apply:
        
        Error: Error creating Kinesis Video Stream: InvalidParameter: 1 validation error(s) found.
        - minimum field size of 1, CreateStreamInput.Tags.
```

To workaround this, the best option is usually to do a length check first:

```go
if v, ok := d.GetOk("tags"); ok && len(v) > 0 {
	createOpts.Tags = keyvaluetags.New(v.(map[string]interface{})).IgnoreAws().KinesisvideoTags()
}
```

Reply:
fixed d734652